File: Low Level Design (LLD)/DesignPatterns/3. Structural/Observer/Java/WeatherStation.java
import java.util.ArrayList;
import java.util.List;

public class WeatherStation implements Subject {
    private List<Observer> observers = new ArrayList<>();
    private float temperature;

    public void setTemperature(float temperature) {
        this.temperature = temperature;
        notifyObservers();  // Notify all observers when temperature changes
    }

    public float getTemperature() {
        return temperature;
    }

    @Override
    public void addObserver(Observer o) {
        observers.add(o);
    }

    @Override
    public void removeObserver(Observer o) {
        observers.remove(o);
    }

    @Override
    public void notifyObservers() {
        for (Observer observer : observers) {
            observer.update(temperature);
        }
    }
}

File: Low Level Design (LLD)/DesignPatterns/3. Structural/Observer/Cpp/src/WeatherStation.cpp
#include "../include/WeatherStation.h"
#include <algorithm>

void WeatherStation::setTemperature(float temp) {
    temperature = temp;
    notifyObservers();
}

float WeatherStation::getTemperature() const {
    return temperature;
}

void WeatherStation::addObserver(Observer* observer) {
    observers.push_back(observer);
}

void WeatherStation::removeObserver(Observer* observer) {
    observers.erase(std::remove(observers.begin(), observers.end(), observer), observers.end());
}

void WeatherStation::notifyObservers() {
    for (Observer* obs : observers) {
        obs->update(temperature);
    }
}
